File: app/src/main/java/ua/com/amicablesoft/android/wr/ui/DrawerAdapter.java
package ua.com.amicablesoft.android.wr.ui;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.RadioButton;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import ua.com.amicablesoft.android.wr.R;
import ua.com.amicablesoft.android.wr.models.Specification;

/**
 * Created by olha on 2/16/17.
 */

class DrawerAdapter extends BaseAdapter {

    interface ButtonClickListener {
        void onButtonClick(Specification specification);
    }

    private static final int TITLE = 0;
    private static final int CHECKBOX = 1;
    private static final int RADIO_BUTTON = 2;
    private static final int BUTTON = 3;
    private static final int COUNT = 4;
    private List<String> items = new ArrayList<>();
    private LayoutInflater layoutInflater;
    private List<RadioButton> radioButtons = new ArrayList<>();
    private ButtonClickListener buttonClickListener;
    private Specification specification;

    DrawerAdapter(Context context) {
        layoutInflater = (LayoutInflater.from(context));
        specification = new Specification();
    }

    @Override
    public int getCount() {
        return items.size();
    }

    @Override
    public Object getItem(int position) {
        return items.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0 || position == 4) {
            return TITLE;
        } else if (position == 1 || position == 2 || position == 3) {
            return CHECKBOX;
        } else if (position == getCount() - 1) {
            return BUTTON;
        } else {
            return RADIO_BUTTON;
        }
    }

    @Override
    public int getViewTypeCount() {
        return COUNT;
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {
        ViewHolder viewHolder = null;
        int type = getItemViewType(position);
        if (convertView == null) {
            viewHolder = new ViewHolder();
            switch (type) {
                case TITLE:
                    convertView = layoutInflater.inflate(R.layout.drawer_title, parent, false);
                    convertView.setEnabled(false);
                    viewHolder.textView = (TextView) convertView.findViewById(R.id.title_text_view);
                    viewHolder.textView.setText(items.get(position));
                    break;
                case CHECKBOX:
                    convertView = layoutInflater.inflate(R.layout.drawer_checkbox, parent, false);
                    viewHolder.checkBox = (CheckBox) convertView.findViewById(R.id.check_box);
                    viewHolder.checkBox.setText(items.get(position));
                    viewHolder.checkBox.setChecked(true);
                    final ViewHolder finalViewHolder1 = viewHolder;
                    viewHolder.checkBox.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            if (finalViewHolder1.checkBox.isChecked()) {
                                if (position == 1) {
                                    specification.setSquats(true);
                                } else if (position == 2) {
                                    specification.setBenchPress(true);
                                } else if (position == 3) {
                                    specification.setDeadLift(true);
                                }
                            } else {
                                if (position == 1) {
                                    specification.setSquats(false);
                                } else if (position == 2) {
                                    specification.setBenchPress(false);
                                } else if (position == 3) {
                                    specification.setDeadLift(false);
                                }
                            }
                        }
                    });
                    break;
                case RADIO_BUTTON:
                    convertView = layoutInflater.inflate(R.layout.drawer_radio_button, parent, false);
                    viewHolder.radioButton = (RadioButton) convertView.findViewById(R.id.radio_button);
                    viewHolder.radioButton.setText(items.get(position));
                    radioButtons.add(viewHolder.radioButton);
                    radioButtons.get(0).setChecked(true);
                    final ViewHolder finalViewHolder = viewHolder;
                    viewHolder.radioButton.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            if (finalViewHolder.radioButton.isChecked()) {
                                for (int i = 0; i < radioButtons.size(); i++) {
                                    RadioButton radioButton = radioButtons.get(i);
                                    if (!radioButton.equals(finalViewHolder.radioButton)) {
                                        radioButton.setChecked(false);
                                    }
                                }
                                specification.setCompetition(finalViewHolder.radioButton.getText().toString());
                            }
                        }
                    });
                    break;
                case BUTTON:
                    convertView = layoutInflater.inflate(R.layout.drawer_button, parent, false);
                    viewHolder.button = (Button) convertView.findViewById(R.id.button);
                    viewHolder.button.setText(items.get(position));
                    viewHolder.button.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            buttonClickListener.onButtonClick(specification);
                        }
                    });
                    break;
            }
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }
        return convertView;
    }

    void setButtonClickListener(ButtonClickListener buttonClickListener) {
        this.buttonClickListener = buttonClickListener;
    }

    void setListItems(List<String> items) {
        this.items.clear();
        this.items.addAll(items);
        notifyDataSetChanged();
    }

    private static class ViewHolder {
        TextView textView;
        CheckBox checkBox;
        RadioButton radioButton;
        Button button;
    }
}

File: app/src/main/java/ua/com/amicablesoft/android/wr/ui/gallery/DrawerAdapter.kt
/*
 * Created by Olha Tymoshenko on 5/24/18 9:22 PM.
 * Copyright (c) 2018. All rights reserved.
 */

package ua.com.amicablesoft.android.wr.ui.gallery

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import ua.com.amicablesoft.android.wr.R
import ua.com.amicablesoft.android.wr.dal.Specification
import java.util.*

class DrawerAdapter(context: Context) : BaseAdapter() {

    companion object {
        private val TITLE = 0
        private val CHECKBOX = 1
        private val RADIO_BUTTON = 2
        private val BUTTON = 3
        private val COUNT = 4
    }

    private val items = ArrayList<String>()
    private val layoutInflater: LayoutInflater = LayoutInflater.from(context)
    private val radioButtons = ArrayList<RadioButton>()
    private var buttonClickListener: ButtonClickListener? = null
    private val specification: Specification = Specification()

    interface ButtonClickListener {
        fun onButtonClick(specification: Specification)
    }

    override fun getCount(): Int {
        return items.size
    }

    override fun getItem(position: Int): Any {
        return items[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0 || position == 4) {
            TITLE
        } else if (position == 1 || position == 2 || position == 3) {
            CHECKBOX
        } else if (position == count - 1) {
            BUTTON
        } else {
            RADIO_BUTTON
        }
    }

    override fun getViewTypeCount(): Int {
        return COUNT
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        val viewHolder: ViewHolder?
        val type = getItemViewType(position)
        if (view == null) {
            viewHolder = ViewHolder()
            when (type) {
                TITLE -> {
                    view = layoutInflater.inflate(R.layout.drawer_title, parent, false)
                    view!!.isEnabled = false
                    viewHolder.textView = view.findViewById<View>(R.id.title_text_view) as TextView
                    viewHolder.textView!!.text = items[position]
                }
                CHECKBOX -> {
                    view = layoutInflater.inflate(R.layout.drawer_checkbox, parent, false)
                    viewHolder.checkBox = view!!.findViewById<View>(R.id.check_box) as CheckBox
                    viewHolder.checkBox!!.text = items[position]
                    viewHolder.checkBox!!.isChecked = true
                    viewHolder.checkBox!!.setOnClickListener {
                        if (viewHolder.checkBox!!.isChecked) {
                            when (position) {
                                1 -> specification.isSquats = true
                                2 -> specification.isBenchPress = true
                                3 -> specification.isDeadLift = true
                            }
                        } else {
                            when (position) {
                                1 -> specification.isSquats = false
                                2 -> specification.isBenchPress = false
                                3 -> specification.isDeadLift = false
                            }
                        }
                    }
                }
                RADIO_BUTTON -> {
                    view = layoutInflater.inflate(R.layout.drawer_radio_button, parent, false)
                    viewHolder.radioButton = view!!.findViewById<View>(R.id.radio_button) as RadioButton
                    viewHolder.radioButton!!.text = items[position]
                    radioButtons.add(viewHolder.radioButton!!)
                    radioButtons[0].isChecked = true
                    val finalViewHolder = viewHolder
                    viewHolder.radioButton!!.setOnClickListener {
                        if (finalViewHolder.radioButton!!.isChecked) {
                            for (i in radioButtons.indices) {
                                val radioButton = radioButtons[i]
                                if (radioButton != finalViewHolder.radioButton) {
                                    radioButton.isChecked = false
                                }
                            }
                            specification.competition = finalViewHolder.radioButton!!.text.toString()
                        }
                    }
                }
                BUTTON -> {
                    view = layoutInflater.inflate(R.layout.drawer_button, parent, false)
                    viewHolder.button = view!!.findViewById<View>(R.id.button) as Button
                    viewHolder.button!!.text = items[position]
                    viewHolder.button!!.setOnClickListener { buttonClickListener!!.onButtonClick(specification) }
                }
            }
            view!!.tag = viewHolder
        } else {
            viewHolder = view.tag as ViewHolder
        }
        return view
    }

    fun setButtonClickListener(buttonClickListener: ButtonClickListener) {
        this.buttonClickListener = buttonClickListener
    }

    fun setListItems(items: List<String>) {
        this.items.clear()
        this.items.addAll(items)
        notifyDataSetChanged()
    }

    private class ViewHolder {
        internal var textView: TextView? = null
        internal var checkBox: CheckBox? = null
        internal var radioButton: RadioButton? = null
        internal var button: Button? = null
    }
}
